components/places/src/storage/bookmarks.rs

```diff
@@ -1236,6 +1236,173 @@ pub fn fetch_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<Book
+        raw
+    } else {
+        return Ok(None);
+    };
+
+    // If we're a folder that has children, fetch child guids or children depending.
+    let (child_guids, child_nodes) =
+        if rb.bookmark_type == BookmarkType::Folder && rb.child_count != 0 {
+            let child_guids: Vec<String> = db.query_rows_into(
+                "SELECT guid
+                 FROM moz_bookmarks
+                 WHERE parent = :parent
+                 ORDER BY position ASC",
+                &[(":parent", &rb.row_id)],
+                |row| row.get_checked(0),
+            )?;
+            if get_direct_children {
+                let children: Vec<_> = child_guids
+                    .into_iter()
+                    .map(|guid_string| {
+                        let child_guid = SyncGuid(guid_string);
+                        if let Some(bmk) = fetch_bookmark_in_tx(db, &child_guid, false)? {
+                            Ok(bmk)
+                        } else {
+                            // Not ideal (since this shouldn't be possible, we're in
+                            // a transaciton, and just fetched these guids), but
+                            // restructuring our queries so that this is impossible
+                            // is tricky, and it seems better to have an error
+                            // that's never actually used than to unwrap()
+                            Err(Error::from(Corruption::MissingChild {
+                                parent: item_guid.0.clone(),
+                                child: child_guid.0,
+                            }))
+                        }
+                    })
+                    .collect::<Result<_>>()?;
+                // Note: even though we have the child guids, we don't return them
+                // because we don't want to send both over the FFI, and the child nodes
+                // should have enough information.
+                (vec![], children)
+            } else {
+                (child_guids, vec![])
+            }
+        } else {
+            (vec![], vec![])
+        };
+
+    let result = ProtoBookmark {
+        node_type: Some(rb.bookmark_type as i32),
+        guid: Some(rb.guid.0),
+        parent_guid: Some(rb.parent_guid.0),
+        position: Some(rb.position),
+        date_added: Some(rb.date_added.0 as i64),
+        last_modified: Some(rb.date_modified.0 as i64),
+        url: rb.url.map(|u| u.into_string()),
+        title: rb.title,
+        child_guids,
+        child_nodes,
+        have_child_nodes: Some(rb.bookmark_type == BookmarkType::Folder && get_direct_children),
+    };
+
+    Ok(Some(result))
+}
+
+/// Call fetch_tree, convert the result to a ProtoBookmark, and ensure the
+/// requested item's position and parent info are provided as well. This is
+/// the function called by the FFI when requesting the tree.
+pub fn fetch_proto_tree(db: &impl ConnExt, item_guid: &SyncGuid) -> Result<Option<ProtoBookmark>> {
```

Comment:
I'm a little uneasy about the proto stuff being in this module - how hard would it be to have that in a peer module?

Reply:
Not hard.